server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoHashCellIdSource.java

```diff
@@ -0,0 +1,108 @@
+import org.elasticsearch.geometry.Rectangle;
+import org.elasticsearch.geometry.utils.Geohash;
+import org.elasticsearch.index.fielddata.MultiGeoPointValues;
+import org.elasticsearch.index.fielddata.SortedBinaryDocValues;
+import org.elasticsearch.index.fielddata.SortedNumericDoubleValues;
+import org.elasticsearch.search.aggregations.support.ValuesSource;
+
+/**
+ * Class to help convert {@link MultiGeoPointValues}
+ * to GeoHash bucketing.
+ */
+public class GeoHashCellIdSource extends ValuesSource.Numeric {
+    private final GeoPoint valuesSource;
+    private final int precision;
+    private final GeoBoundingBox geoBoundingBox;
+
+    public GeoHashCellIdSource(GeoPoint valuesSource, int precision, GeoBoundingBox geoBoundingBox) {
+        this.valuesSource = valuesSource;
+        this.precision = precision;
+        this.geoBoundingBox = geoBoundingBox;
+    }
+
+    public int precision() {
+        return precision;
+    }
+
+    @Override
+    public boolean isFloatingPoint() {
+        return false;
+    }
+
+    @Override
+    public SortedNumericDocValues longValues(LeafReaderContext ctx) {
+        return geoBoundingBox.isUnbounded() ?
+            new UnboundedCellValues(valuesSource.geoPointValues(ctx), precision) :
+            new BoundedCellValues(valuesSource.geoPointValues(ctx), precision, geoBoundingBox);
+    }
+
+    @Override
+    public SortedNumericDoubleValues doubleValues(LeafReaderContext ctx) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public SortedBinaryDocValues bytesValues(LeafReaderContext ctx) {
+        throw new UnsupportedOperationException();
+    }
+
+    private static class UnboundedCellValues extends CellValues {
+
+        UnboundedCellValues(MultiGeoPointValues geoValues, int precision) {
+            super(geoValues, precision);
+        }
+
+        @Override
+        int advanceValue(org.elasticsearch.common.geo.GeoPoint target, int valuesIdx) {
+            values[valuesIdx] = Geohash.longEncode(target.getLon(), target.getLat(), precision);
+            return valuesIdx + 1;
+        }
+    }
+
+    private static class BoundedCellValues extends CellValues {
+
+        private final GeoBoundingBox bbox;
+        private final boolean crossesDateline;
+
+        BoundedCellValues(MultiGeoPointValues geoValues, int precision, GeoBoundingBox bbox) {
+            super(geoValues, precision);
+            this.bbox = bbox;
+            this.crossesDateline = bbox.right() < bbox.left();
+        }
+
+        @Override
+        int advanceValue(org.elasticsearch.common.geo.GeoPoint target, int valuesIdx) {
+            final String hash = Geohash.stringEncode(target.getLon(), target.getLat(), precision);
+            if (validHash(hash)) {
+                values[valuesIdx] = Geohash.longEncode(hash);
+                return valuesIdx + 1;
+            }
+            return valuesIdx;
+        }
+
+        private boolean validHash(String hash) {
```

Comment:
I have seen this one somewhere else already. Time to move it to org.elasticsearch.common.geo.GeoUtils?

Reply:
I don't think it fits in GeoUtils nicely. I hope this duplication will go when/if we move GeoShapeWithdocValues to server.